Added Support for TSQL's column_privileges

oscarno22 opened the pull request:
### Description

Currently Babelfish does not support
information_schema_tsql.column_privileges. This commit updates this
state by implementing the view.

TASK:

Author: Ozzy Nolen <oscarno@amazon.com>
Signed-off by:
 
### Issues Resolved

- The view information_schema_tsql.column_privileges not implemented


### Check List
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

oscarno22 committed:
Added Support for TSQL's column_privileges

Currently Babelfish does not support
information_schema_tsql.column_privileges. This commit updates this
state by implementing the view.

TASK:

Signed-off-by: Ozzy Nolen <oscarno@amazon.com>

oscarno22 committed:
Updated Views to TSQL Syntax

This commit converts the view definition to TSQL syntax, especially as
it pertains to CAST().

TASK:

Signed-off-by: Ozzy Nolen <oscarno@amazon.com>

timchang514 commented:
Is there a JIRA that is linked to this issue?

timchang514 commented on contrib/babelfishpg_tsql/sql/information_schema_tsql.sql:
```diff
@@ -717,3 +717,73 @@ CREATE OR REPLACE VIEW information_schema_tsql.routines AS
+            pr_c.prtype,
+            pr_c.grantable,
+            pr_c.relowner
+           FROM ( SELECT pg_class.oid,
```
do you have any performance numbers on this? could you include a comparison of how column_privileges performs here compared to in sql server?

oscarno22 replied:
Currently, the interns and myself do not have performance numbers. What tools do we need to get this?

oscarno22 replied:
I ran the testing files against Client Statistics in SSMS. They ran in 2932 ms in Babelfish and around 2600 in sql server.

xhfanhe commented on test/JDBC/input/column_privileges-vu-prepare.sql:
```diff
@@ -0,0 +1,35 @@
+create login column_privileges_vu_prepare_log with password = 'EmptyPassword#';
+go
+
+create user column_privileges_vu_prepare_user for login column_privileges_vu_prepare_log;
```
CREATE USER and GRANT statement are not supported until Babelfish 13.6. I'll suggest move this part to verify file.

oscarno22 replied:
Resolved in latest commit.

xhfanhe commented on test/JDBC/input/column_privileges-vu-prepare.sql:
```diff
@@ -0,0 +1,35 @@
+use master;
```
We should prepare one view, one procedure and one function that depend on  information_schema_tsql.column_privileges

oscarno22 replied:
Resolved in latest commit.

xhfanhe commented on contrib/babelfishpg_tsql/sql/information_schema_tsql.sql:
```diff
@@ -717,3 +717,73 @@ CREATE OR REPLACE VIEW information_schema_tsql.routines AS
+*/
+
+CREATE OR REPLACE VIEW information_schema_tsql.column_privileges
+AS SELECT CAST(u_grantor.rolname AS sys.nvarchar(128)) AS "GRANTOR",
```
GRANTOR pg_roles.rolname is showing the physical PG role name, but we want the logical role name (i.e. the mapped name for TSQL) here, which means `orig_username` stored in `sys.babelfish_authid_user_ext`.
The same applies to GRANTEE and TABLE_SCHEMA nspname. The latter should be selected from `sys.babelfish_namespace_ext.orig_name`

oscarno22 replied:
Will be updated in next commit.

oscarno22 replied:
Resolved in latest commit.

oscarno22 committed:
Updated Testing Files

This commit updates testing files to account for functionality and
dependency. ISC column_privileges is called with the creation of tables and verified
with a view, function, and procedure that depend on it.

TASK:

Signed-off-by: Ozzy Nolen <oscarno@amazon.com>

oscarno22 committed:
Updated Naming in Testing Files

This commit updates the view, function, and procedure created in the
verify file to follow the naming practice:
"column_privileges_vu_verify_%".

TASK:

Signed-off-by: Ozzy Nolen <oscarno@amazon.com>

oscarno22 committed:
Removed Custom Database from Test Files

This commit removes the custom testing database and opts to use the one
created in the test framework.

TASK:

Signed-off-by: Ozzy Nolen <oscarno@amazon.com>

oscarno22 committed:
Updates Columns to Retrieve Logical Role Name

Previously, the view definition pulled the physical pg role name for
grantor, grantee, and table_schema. This commit retrieves the logical
role name from sys.babelfish_authid_username_ext and
sys.babelfish_namespace_ext.

TASK:

Signed-off-by: Ozzy Nolen <oscarno@amazon.com>

xhfanhe reviewed (changes requested):
Code LGTM.
Version upgrade test [source_latest-target_latest] is failing with the following error
```
--- /home/runner/work/babelfish_extensions/babelfish_extensions/test/python/expected/upgrade_validation/expected_dependency.out	2022-08-12 21:37:21.093776744 +0000
+++ /home/runner/work/babelfish_extensions/babelfish_extensions/test/python/output/upgrade_validation/expected_dependency.out	2022-08-12 21:49:47.536206914 +0000
@@ -1007,6 +1007,7 @@
 Type sys.rowversion
 View information_schema_tsql.check_constraints
 View information_schema_tsql.column_domain_usage
+View information_schema_tsql.column_privileges
 View information_schema_tsql.columns
 View information_schema_tsql.constraint_column_usage
 View information_schema_tsql.domains

2022-08-12 21:49:47,546-ERROR-Test Failed!
```
Please update `babelfish_extensions/test/python/expected/upgrade_validation/expected_dependency.out` this file accordingly

oscarno22 committed:
Updated File for Version Upgrade Tests

Updated expected_dependency.out file to account for
information_schema_tsql.column_privileges.

TASK:

Signed-off-by: Ozzy Nolen <oscarno@amazon.com>

oscarno22 committed:
Added Privilege Test File

This commits adds a testing file specifically for the select privilege. A
.mix file was used in order to switch between different user contexts. A
custom database was used so that role objects would not become
duplicated in the master database.

TASK:

Signed-off-by: Ozzy Nolen <oscarno@amazon.com>

oscarno22 committed:
Added Database Test File

This commit adds an additional test file checking that the view does not
retrieve objects across databases.

TASK:

Signed-off-by: Ozzy Nolen <oscarno@amazon.com>

oscarno22 committed:
Updated View Definition and Database Test File

This commit replaces pg_namespace with sys.pg_namespace_ext joined by
babelfish_namespace_ext. Grantee is also joined by
sys.babelfish_authid_user_ext.  Also, instead of calling sys.db_name()
directly, the query result is now conditioned. Finally, the database
test file not accounts for a schema with its own table.

TASK:

Signed-off-by: Ozzy Nolen <oscarno@amazon.com>
